File: LinkedList/src/LinkedList.java
public class LinkedList<E> {

    private class Node{
        public E e;
        public Node next;

        public Node(E e, Node next){
            this.e = e;
            this.next = next;
        }

        public Node(E e){
            this(e, null);
        }

        public Node(){
            this(null, null);
        }

        @Override
        public String toString(){
            return e.toString();
        }
    }

    private Node dummyhead;   // 添加一个虚拟头节点
    int size;

    public LinkedList(){
        dummyhead = new Node(null, null);
        size = 0;
    }

    // 获取链表元素个数
    public int getSize(){
        return size;
    }

    // 判断链表是否为空
    public boolean isEmpty(){
        return size == 0;
    }


    // 在链表index处添加节点
    // 关键是找到前一个节点
    public void add(int index, E e){
        if (index < 0 || index > size) {
            throw new IllegalArgumentException("failed");
        }

//        if (index == 0){
//            addFirst(e);
//        }
//        else{
//            Node prev = head;
//            for (int i = 0 ; i < index -1; i ++){
//                prev = prev.next;
//            }
////            Node node = new Node(e);
////            node.next = prev.next;
////            prev.next = node;
//            prev.next = new Node(e, prev.next);
//            size ++;
//        }

        Node prev = dummyhead;
        for (int i = 0 ; i < index; i ++){
            prev = prev.next;   // 找插入节点的前一个节点
        }
        prev.next = new Node(e, prev.next);
        size ++;
    }

    //在链表头部添加新元素
    public void addFirst(E e){
//        Node node = new Node(e);
//        node.next = head;
//        head = node;
//        head = new Node(e, head);
//        size ++;
        add(0, e);
    }

    public void addLast(E e){
        add(size, e);
    }


    // 获得链表第index处的元素
    public E get(int index){
        if (index < 0 || index > size) {
            throw new IllegalArgumentException("failed");
        }

        // 遍历元素
        Node cur = dummyhead.next;
        for (int i = 0; i < index; i ++){
            cur = cur.next;
        }
        return cur.e;
    }

    // 获得第一个元素
    public E getFist(){
        return get(0);
    }
    // 获得最后一个元素
    public E getLast(){
        return get(size - 1);
    }

    // 修改index处的元素
    public void set(int index, E e){
        if (index < 0 || index > size) {
            throw new IllegalArgumentException("failed");
        }

        Node cur = dummyhead.next;
        for (int i = 0 ; i < index; i ++){
            cur = cur.next;
        }
        cur.e = e;
    }

    // 如果是基本类型比较，那么只能用==来比较，不能用equals ，如果是基本类型的包装类型,那么用equals
    // 查找数据
    public boolean contains(E e){
        Node cur = dummyhead.next;
        while (cur != null){
            if (cur.e.equals(e)){
                return true;
            }
            cur = cur.next;
        }
        return false;
    }

    // 从index位置删除一个元素
    public E remove(int index){
        if (index < 0 || index > size) {
            throw new IllegalArgumentException("failed");
        }
        Node prev = dummyhead;
        for (int i = 0 ; i < index; i ++){
            prev = prev.next;   // 找到待删除节点的前一个节点
        }

        Node retNode = prev.next;
        prev.next = retNode.next;
        retNode.next = null;
        size --;

        return retNode.e;

    }

    public E removeFist(){
        return remove(0);
    }

    public E removeLast(){
        return remove(size - 1);
    }

    @Override
    public String toString(){
        StringBuilder res = new StringBuilder();

        Node cur = dummyhead.next;
        while(cur != null){
            res.append(cur + " -> ");
            cur = cur.next;
        }
//        for (Node cur = dummyhead.next; cur != null; cur = cur.next)
//            res.append(cur + " -> ");
        res.append("NULL");
        return res.toString();
    }
}

File: _implement/linkedlist.c
//
// Created by hpcomputer on 2019/11/30.
//

//
// Created by hpcomputer on 2019/11/30.
//

#include<stdio.h>
#include<stdlib.h>


typedef struct linked_list{
    int data;
    struct linked_list *next;
}linked_list_node, *linked_list;


// 初始化创建虚拟头节点,数据从1开始计数
void init_linked_list(linked_list *l){
    *l = malloc(sizeof(linked_list_node));
    (*l) -> next = NULL;
    // data可以存任何或节点个数
    // printf("l is addr: %d\n", *l);
}


// 判断链表是否为空
int isEmpty(linked_list l){
    return l.next == NULL;
}

// 获取链表元素个数
int get_size(linked_list l){
    if (l -> next == NULL){
        return 0;
    }
    linked_list_node *p;
    int count = 0;
    for ( p = l -> next; p -> next != NULL; p = p -> next){
        count += 1;
        // printf("count : %d\n", count);
    }
    return count;

}

// 在链表index处添加节点
void add_index(linked_list *l, int index, int e){
    linked_list_node *prev;
    prev = (*l);
    for (int i = 0; i < index && prev -> next !=NULL; i ++){
        prev = prev -> next;
        // printf("prev -> data = %d\n", prev -> data);
    }
    linked_list_node *p = malloc(sizeof(linked_list_node));
    p -> data = e;
    p -> next = prev -> next;
    prev -> next = p;
}

// 在链表头部添加节点、
void add_first(linked_list *l, int e){
    add_index(l, -1, e);
}

// 在链表尾部添加节点
void add_last(linked_list *l, int e){
    // linked_list_node *prev;
    // rev = (*l);
    // for ( ; prev -> next != NULL; ){
    //     prev = prev -> next;
    //     // printf("prev -> data = %d\n", prev -> data);
    // }

    // // for (int i = 0 ; i < index)
    // linked_list_node *p = malloc(sizeof(linked_list_node));
    // p -> data = e;
    // p -> next = prev -> next;
    // prev -> next = p;p
    add_index(l, get_size(*l)+1, e);
}

// 获取链表index处的数据
int get(linked_list l, int index){
    linked_list_node *p = l -> next;
    for (int i = 0; i < index; i++){
        p = p -> next;
    }
    return p -> data;
}

// // 获取第一个数据
int get_first(linked_list l){
    return get(l, 0);
}

// // 获取最后一个元素
int get_last(linked_list l){
    return get(l, get_size(l));
}

// 查找数据
int contains(linked_list l, int e){
    linked_list_node *p = l -> next;
    int count = 1;
    while (p != NULL){
        if (p -> data = e){
            return 1;
        }
        p = p -> next;
    }
    return 0;
}


// 移除第index位置上的数据
int remove_index(linked_list *l, int index){
    linked_list_node *prev;
    prev = (*l) -> next;   // 由于该链表存在虚拟头节点，则，循环从1开始
    for (int i = 1; i <= index; i ++){
        // printf("prev -> data = %d\n", prev ->data);
        prev = prev -> next;
    }

    linked_list_node *ret = prev -> next;
    prev -> next = ret -> next;
    return ret -> data;
}

int remove_first(linked_list *l){
    return remove_index(l, 1);
}

int remove_last(linked_list *l){
    return remove_index(l, get_size(*l));
}


// 移除特定的数据
void remove_element(linked_list *l, int val){
    printf("remove_element\n");
    linked_list_node *p;
    p = (*l);
    int count = 0;
    while (p -> next != NULL){
        // printf("p -> data = %d\n", p -> data);
        if (p -> next -> data == val){
            p -> next = p -> next -> next;
            count += 1;
        }
        p = p -> next;
    }

}

//// 利用递归实现
void remove_elements_circulation(linked_list *L, int x)
{
    if ((*L) == NULL)
        return;

    linked_list_node *t;
    if ((*L) -> data == x){
        t = *L;
        (*L) = (*L) -> next;
        free(t);
        return;
    }
    else
        remove_elements_circulation(&((*L) -> next) , x);
}

// 打印链表
void print_linked_list(linked_list *l){
    printf("print linkedlist\n");
    linked_list_node *p = (*l) -> next;
    int count = 1;
    while (p != NULL){
        printf("%d(%d) -> ", p -> data, count);
        p = p -> next;
        count += 1;
    }
    printf("NULL\n");
}

int main(void)
{
    /* code */
    linked_list l;
    // printf("start\n");
    init_linked_list(&l);
    // printf("init end ");
    for (int i = 0; i <= 10; i ++){
        add_last(&l, i);
    }
    // printf("get_size = %d\n", get_size(l));
    // print_linked_list(&l);
    // printf("contains = 0 is %d\n", contains(l,0));
    // printf("get(3) = %d\n", get(l, 3));
    // int e;
    // e = remove_index(&l, 2);
    // printf("remove_index(&l, 4) = \n", e);
//    print_linked_list(&l);

//    remove_element(&l, 2);
    remove_elements_circulation(&l, 7);
    print_linked_list(&l);
    printf("hello world\n");
    return 0;
}